File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/color-compatibility/gaussian-blur-with-mask.ts
//based on thread https://forum.unity.com/threads/contribution-texture2d-blur-in-c.185694/

import { Color } from '../../../core/models/color.model';
import { Size } from '../../../core/models/size.interface';
import { Vector } from '../../../core/models/vector.model';

class ColorSum {
    private r: number;
    private g: number;
    private b: number;
    private a: number;

    private count: number;

    public add(color: Color): void {
        this.r += color.r;
        this.g += color.g;
        this.b += color.b;
        this.a += color.a ?? 0;

        this.count++;
    }

    public subtract(color: Color): void {
        this.r -= color.r;
        this.g -= color.g;
        this.b -= color.b;
        this.a -= color.a;

        this.count--;
    }

    public GetColor(): Color {
        return {
            r: (this.r / this.count) * 255,
            g: (this.g / this.count) * 255,
            b: (this.b / this.count) * 255,
            a: (this.a / this.count) * 255,
        };
    }
}

export class GaussianBlurWithMask {
    private readonly oryginalColors: Color[];
    private readonly width: number;
    private readonly height: number;

    private mask: boolean[];

    public constructor(oryginalColors: Color[], size: Size) {
        this.oryginalColors = oryginalColors;
        this.width = size.width;
        this.height = size.height;
    }

    public blur(mask: boolean[], radius: number, iterations: number): Color[] {
        this.mask = mask;

        const arrayA: Color[] = [...this.oryginalColors];
        const arrayB: Color[] = new Array(this.oryginalColors.length);

        for (let i = 0; i < iterations; ++i) {
            this.horizontalBlur(arrayA, arrayB, radius);
            this.verticalBlur(arrayB, arrayA, radius);
        }

        return arrayA;
    }

    private horizontalBlur(inputColors: Color[], outputColors: Color[], radius: number): void {
        //paralele byli tu
        for (let imgY = 0; imgY < this.height; ++imgY) {
            const colorSum: ColorSum = new ColorSum();

            for (let imgX = 0; imgX < this.width; ++imgX) {
                if (imgX == 0) {
                    for (let x = 0; x <= radius; ++x) {
                        const pos: Vector = new Vector(x, imgY);
                        if (this.isPixelInArrayAndIsVisable(pos)) {
                            colorSum.add(inputColors[pos.y * this.width + pos.x]);
                        }
                    }
                } else {
                    const pxToSubstractPos: Vector = new Vector(imgX - radius - 1, imgY);
                    if (this.isPixelInArrayAndIsVisable(pxToSubstractPos)) {
                        colorSum.subtract(inputColors[pxToSubstractPos.y * this.width + pxToSubstractPos.x]);
                    }

                    const pxToAddPos: Vector = new Vector(imgX + radius, imgY);
                    if (this.isPixelInArrayAndIsVisable(pxToAddPos)) {
                        colorSum.add(inputColors[pxToAddPos.y * this.width + pxToAddPos.x]);
                    }
                }

                const index: number = imgY * this.width + imgX;
                outputColors[index] = this.mask[index] ? colorSum.GetColor() : inputColors[index];
            }
        }
    }

    private verticalBlur(inputColors: Color[], outputColors: Color[], radius: number): void {
        for (let imgX = 0; imgX < this.width; ++imgX) {
            const colorSum: ColorSum = new ColorSum();

            for (let imgY = 0; imgY < this.height; ++imgY) {
                if (imgY == 0) {
                    for (let y = -radius; y <= radius; ++y) {
                        const pos: Vector = new Vector(imgX, y);
                        if (this.isPixelInArrayAndIsVisable(pos)) {
                            colorSum.add(inputColors[pos.y * this.width + pos.x]);
                        }
                    }
                } else {
                    const pxToSubstractPos: Vector = new Vector(imgX, imgY - radius - 1);
                    if (this.isPixelInArrayAndIsVisable(pxToSubstractPos)) {
                        colorSum.subtract(inputColors[pxToSubstractPos.y * this.width + pxToSubstractPos.x]);
                    }

                    const pxToAddPos: Vector = new Vector(imgX, imgY + radius);
                    if (this.isPixelInArrayAndIsVisable(pxToAddPos)) {
                        colorSum.add(inputColors[pxToAddPos.y * this.width + pxToAddPos.x]);
                    }
                }

                const index: number = imgY * this.width + imgX;
                outputColors[index] = this.mask[index] ? colorSum.GetColor() : inputColors[index];
            }
        }
    }

    private isPixelInArrayAndIsVisable(point: Vector): boolean {
        if (point.x < 0 || point.x >= this.width || point.y < 0 || point.y >= this.height) {
            return false;
        }

        return this.mask[point.y * this.width + point.x];
    }
}

File: backend/Irrtirer/Irrtirer.Generator/ColorCompatibility/GaussianBlurWithMask.cs
//based on thread https://forum.unity.com/threads/contribution-texture2d-blur-in-c.185694/

using System.Drawing;
using System.Threading.Tasks;

namespace Irrtirer.Generator.ColorCompatibility
{
    internal class GaussianBlurWithMask
    {
        private struct ColorSum
        {
            private int r;
            private int g;
            private int b;
            private int a;

            private int count;

            public void Add(Color color)
            {
                r += color.R;
                g += color.G;
                b += color.B;
                a += color.A;

                count++;
            }

            public void Subtract(Color color)
            {
                r -= color.R;
                g -= color.G;
                b -= color.B;
                a -= color.A;

                count--;
            }

            public Color GetColor()
            {
                return Color.FromArgb(a / count, r / count, g / count, b / count);
            }
        }

        private readonly Color[] oryginalColors;
        private readonly int width;
        private readonly int height;

        private bool[] mask;

        private readonly ParallelOptions parallelOptions = new ParallelOptions
        {
            MaxDegreeOfParallelism = 8
        };

        public GaussianBlurWithMask(Color[] oryginalColors, Size size)
        {
            this.oryginalColors = oryginalColors;
            width = size.Width;
            height = size.Height;
        }

        public Color[] Blur(bool[] mask, int radius, int iterations)
        {
            this.mask = mask;

            Color[] arrayA = (Color[])oryginalColors.Clone();
            Color[] arrayB = new Color[oryginalColors.Length];

            for (var i = 0; i < iterations; ++i)
            {
                HorizontalBlur(arrayA, arrayB, radius);
                VerticalBlur(arrayB, arrayA, radius);
            }

            return arrayA;
        }

        private void HorizontalBlur(Color[] inputColors, Color[] outputColors, int radius)
        {
            Parallel.For(0, height, parallelOptions, (imgY) =>
            {
                ColorSum colorSum = new ColorSum();

                for (int imgX = 0; imgX < width; ++imgX)
                {
                    if (imgX == 0)
                    {
                        for (int x = 0; x <= radius; ++x)
                        {
                            Point pos = new Point(x, imgY);
                            if (IsPixelInArrayAndIsVisable(pos))
                            {
                                colorSum.Add(inputColors[pos.Y * width + pos.X]);
                            }
                        }
                    }
                    else
                    {
                        Point pxToSubstractPos = new Point(imgX - radius - 1, imgY);
                        if (IsPixelInArrayAndIsVisable(pxToSubstractPos))
                        {
                            colorSum.Subtract(inputColors[pxToSubstractPos.Y * width + pxToSubstractPos.X]);
                        }

                        Point pxToAddPos = new Point(imgX + radius, imgY);
                        if (IsPixelInArrayAndIsVisable(pxToAddPos))
                        {
                            colorSum.Add(inputColors[pxToAddPos.Y * width + pxToAddPos.X]);
                        }
                    }

                    int index = imgY * width + imgX;
                    outputColors[index] = mask[index] ? colorSum.GetColor() : inputColors[index];
                }
            });
        }

        private void VerticalBlur(Color[] inputColors, Color[] outputColors, int radius)
        {
            Parallel.For(0, width, parallelOptions, (imgX) =>
            {
                ColorSum colorSum = new ColorSum();

                for (int imgY = 0; imgY < height; ++imgY)
                {
                    if (imgY == 0)
                    {
                        for (int y = -radius; y <= radius; ++y)
                        {
                            Point pos = new Point(imgX, y);
                            if (IsPixelInArrayAndIsVisable(pos))
                            {
                                colorSum.Add(inputColors[pos.Y * width + pos.X]);
                            }
                        }
                    }
                    else
                    {
                        Point pxToSubstractPos = new Point(imgX, imgY - radius - 1);
                        if (IsPixelInArrayAndIsVisable(pxToSubstractPos))
                        {
                            colorSum.Subtract(inputColors[pxToSubstractPos.Y * width + pxToSubstractPos.X]);
                        }

                        Point pxToAddPos = new Point(imgX, imgY + radius);
                        if (IsPixelInArrayAndIsVisable(pxToAddPos))
                        {
                            colorSum.Add(inputColors[pxToAddPos.Y * width + pxToAddPos.X]);
                        }
                    }

                    int index = imgY * width + imgX;
                    outputColors[index] = mask[index] ? colorSum.GetColor() : inputColors[index];
                }
            });
        }

        private bool IsPixelInArrayAndIsVisable(Point point)
        {
            if (point.X < 0 || point.X >= width || point.Y < 0 || point.Y >= height)
                return false;

            return mask[point.Y * width + point.X];
        }
    }
}
